crates/uv-resolver/src/iterators.rs

```diff
@@ -0,0 +1,104 @@
+use std::cmp::Reverse;
+use std::collections::BinaryHeap;
+use std::iter::Peekable;
+
+use pep440_rs::Version;
+
+use crate::version_map::VersionMapDistHandle;
+
+/// An iterator that returns the maximum version from a set of iterators.
+pub(crate) struct MaxIterator<'a, T: Iterator<Item = (&'a Version, VersionMapDistHandle<'a>)>> {
```

Comment:
What about:

```rust
/// An iterator that returns the maximum version from a set of iterators.
pub(crate) struct MaxIterator<'a> {
    heap: Box<dyn Iterator<Item = (&'a Version, VersionMapDistHandle<'a>)> + 'a>,
}

impl<'a> MaxIterator<'a> {
    pub(crate) fn new<T: Iterator<Item = (&'a Version, VersionMapDistHandle<'a>)> + 'a>(
        iterators: Vec<T>,
    ) -> Self {
        let mut heap: Box<dyn Iterator<Item = (&'a Version, VersionMapDistHandle<'a>)>> =
            Box::new(iter::empty());

        // TODO(konsti): The rev is the correct sorting order, isn't it?
        for iterator in iterators.into_iter().rev() {
            heap = Box::new(heap.merge_by(iterator, |a, b| a.0 >= b.0));
        }

        Self { heap }
    }
}

impl<'a> Iterator for MaxIterator<'a> {
    type Item = (&'a Version, VersionMapDistHandle<'a>);

    fn next(&mut self) -> Option<Self::Item> {
        self.heap.next()
    }
}
```

That's fast even in the pathological boto3 case:
```
Benchmark 1: target/profiling/uv-branch pip compile scripts/requirements/boto3.in --index-strategy unsafe-closest-match
  Time (mean ± σ):     391.2 ms ±   2.3 ms    [User: 350.0 ms, System: 53.1 ms]
  Range (min … max):   388.2 ms … 395.5 ms    10 runs
 
Benchmark 2: target/profiling/uv pip compile scripts/requirements/boto3.in --index-strategy unsafe-closest-match
  Time (mean ± σ):     357.9 ms ±   0.9 ms    [User: 324.2 ms, System: 44.9 ms]
  Range (min … max):   356.4 ms … 359.0 ms    10 runs
 
Benchmark 3: target/profiling/uv pip compile scripts/requirements/boto3.in
  Time (mean ± σ):     354.6 ms ±   1.3 ms    [User: 320.7 ms, System: 45.7 ms]
  Range (min … max):   353.3 ms … 356.6 ms    10 runs
 
Summary
  target/profiling/uv pip compile scripts/requirements/boto3.in ran
    1.01 ± 0.00 times faster than target/profiling/uv pip compile scripts/requirements/boto3.in --index-strategy unsafe-closest-match
    1.10 ± 0.01 times faster than target/profiling/uv-branch pip compile scripts/requirements/boto3.in --index-strategy unsafe-closest-match
```

I'm not sure how readable you consider this, but we can also inline max iterator as:
```rust
iterators.into_iter().rev().fold(
    Box::new(iter::empty()) as Box<dyn Iterator<Item = _>>,
    |acc, iterator| Box::new(acc.merge_by(iterator, |a, b| a.0 >= b.0)),
)
```

Reply:
I tried these benchmarks with extra index urls but got resolver or build errors, so the benchmark is a bit limited

Reply:
I think `next` in that case would be `O(N)` in the number of indexes rather than `O(log(N))` as with the binary heap, right? I guess it's a small constant factor in each case.

Reply:
The number of indexes is typically two (if you're using the setting), and ~always less than ten I would guess, so maybe it doesn't matter.

Reply:
I'd expect peeking and comparing versions is faster than the allocations and memcpy in a heap.

For the theory, the head building is expected `O(num versions total)` to build, `O(num versions total * log(num versions total))` worst case, the peeking is also worst case of `O(num versions total * num indexes)`.

We can try `kmerge_by` and benchmark too if we want.

Reply:
Sure, `kmerge_by` looks most straightforward to me since we don't need any `Box` or `dyn`, only allocates one vec upfront, etc.

But if we benchmark, shouldn't we include at least one other index...?

Reply:
Some quick benchmarks with more indexes, not sure why they are so noisy:

```
Benchmark 1: target/profiling/uv-pr pip compile scripts/requirements/jupyter.in
  Time (mean ± σ):      14.7 ms ±   0.7 ms    [User: 13.0 ms, System: 18.2 ms]
  Range (min … max):    13.5 ms …  15.8 ms    10 runs
 
Benchmark 2: target/profiling/uv-pr pip compile scripts/requirements/jupyter.in --index-strategy unsafe-closest-match --extra-index-url https://test.pypi.org/simple --extra-index-url https://download.pytorch.org/whl/cu118
  Time (mean ± σ):      9.953 s ± 13.549 s    [User: 0.041 s, System: 0.043 s]
  Range (min … max):    2.838 s … 41.590 s    10 runs
 
  Warning: Statistical outliers were detected. Consider re-running this benchmark on a quiet system without any interferences from other programs. It might help to use the '--warmup' or '--prepare' options.
 
Benchmark 3: target/profiling/uv-mine pip compile scripts/requirements/jupyter.in --index-strategy unsafe-closest-match --extra-index-url https://test.pypi.org/simple --extra-index-url https://download.pytorch.org/whl/cu118
  Time (mean ± σ):      9.418 s ± 14.120 s    [User: 0.041 s, System: 0.044 s]
  Range (min … max):    1.626 s … 47.047 s    10 runs
 
  Warning: The first benchmarking run for this command was significantly slower than the rest (47.047 s). This could be caused by (filesystem) caches that were not filled until after the first run. You are already using the '--warmup' option which helps to fill these caches before the actual benchmark. You can either try to increase the warmup count further or re-run this benchmark on a quiet system in case it was a random outlier. Alternatively, consider using the '--prepare' option to clear the caches before each timing run.
 
Summary
  target/profiling/uv-pr pip compile scripts/requirements/jupyter.in ran
  640.31 ± 960.55 times faster than target/profiling/uv-mine pip compile scripts/requirements/jupyter.in --index-strategy unsafe-closest-match --extra-index-url https://test.pypi.org/simple --extra-index-url https://download.pytorch.org/whl/cu118
  676.71 ± 921.83 times faster than target/profiling/uv-pr pip compile scripts/requirements/jupyter.in --index-strategy unsafe-closest-match --extra-index-url https://test.pypi.org/simple --extra-index-url https://download.pytorch.org/whl/cu118
```  

```
Benchmark 1: target/profiling/uv-pr pip compile scripts/requirements/airflow.in
  Time (mean ± σ):     218.6 ms ±   2.6 ms    [User: 205.9 ms, System: 151.8 ms]
  Range (min … max):   214.4 ms … 223.3 ms    10 runs
 
Benchmark 2: target/profiling/uv-pr pip compile scripts/requirements/airflow.in --index-strategy unsafe-closest-match --extra-index-url https://test.pypi.org/simple --extra-index-url https://download.pytorch.org/whl/cu118
  Time (mean ± σ):      8.727 s ±  1.060 s    [User: 0.394 s, System: 0.223 s]
  Range (min … max):    7.264 s … 10.827 s    10 runs
 
Benchmark 3: target/profiling/uv-mine pip compile scripts/requirements/airflow.in --index-strategy unsafe-closest-match --extra-index-url https://test.pypi.org/simple --extra-index-url https://download.pytorch.org/whl/cu118
  Time (mean ± σ):      9.036 s ±  1.217 s    [User: 0.378 s, System: 0.235 s]
  Range (min … max):    7.355 s … 11.375 s    10 runs
 
Summary
  target/profiling/uv-pr pip compile scripts/requirements/airflow.in ran
   39.91 ± 4.87 times faster than target/profiling/uv-pr pip compile scripts/requirements/airflow.in --index-strategy unsafe-closest-match --extra-index-url https://test.pypi.org/simple --extra-index-url https://download.pytorch.org/whl/cu118
   41.32 ± 5.59 times faster than target/profiling/uv-mine pip compile scripts/requirements/airflow.in --index-strategy unsafe-closest-match --extra-index-url https://test.pypi.org/simple --extra-index-url https://download.pytorch.org/whl/cu118
```

```
Benchmark 1: target/profiling/uv-pr pip compile scripts/requirements/boto3.in
  Time (mean ± σ):     360.4 ms ±   0.8 ms    [User: 321.0 ms, System: 50.9 ms]
  Range (min … max):   359.4 ms … 362.3 ms    10 runs
 
Benchmark 2: target/profiling/uv-pr pip compile scripts/requirements/boto3.in --index-strategy unsafe-closest-match --extra-index-url https://test.pypi.org/simple --extra-index-url https://download.pytorch.org/whl/cu118
  Time (mean ± σ):      1.558 s ±  0.580 s    [User: 0.361 s, System: 0.052 s]
  Range (min … max):    0.840 s …  2.441 s    10 runs
 
Benchmark 3: target/profiling/uv-mine pip compile scripts/requirements/boto3.in --index-strategy unsafe-closest-match --extra-index-url https://test.pypi.org/simple --extra-index-url https://download.pytorch.org/whl/cu118
  Time (mean ± σ):      1.750 s ±  0.538 s    [User: 0.363 s, System: 0.052 s]
  Range (min … max):    0.921 s …  2.470 s    10 runs
 
Summary
  target/profiling/uv-pr pip compile scripts/requirements/boto3.in ran
    4.32 ± 1.61 times faster than target/profiling/uv-pr pip compile scripts/requirements/boto3.in --index-strategy unsafe-closest-match --extra-index-url https://test.pypi.org/simple --extra-index-url https://download.pytorch.org/whl/cu118
    4.86 ± 1.49 times faster than target/profiling/uv-mine pip compile scripts/requirements/boto3.in --index-strategy unsafe-closest-match --extra-index-url https://test.pypi.org/simple --extra-index-url https://download.pytorch.org/whl/cu118
```

Reply:
I'm not sure if it's worth optimizing this case to perfection, i'm fine with merging with whatever is easiest

Reply:
I think you need _way_ more than 10 runs and a large warmup for this kind of thing. But I'll just merge whatever looks easiest in the end.